### PR TITLE
388 feature external handlers

### PR DESCRIPTION
**Pull request by dushyantb**
get_external_balance & get_block_timestamp

### CODE REVIEW

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs [hunk 1]
```diff
@@ -378,6 +378,53 @@ public static void Handler_Env_GetMsgValue(int dataOffset)
             if (!ret)
                 throw new InvalidContractException("Bad call to (get_msg_value)");
         }
+        
+        public static void Handler_Env_GetExternalBalance(int addressOffset, int resultOffset)
+        {
+            Logger.LogInformation($"Handler_Env_GetExternalBalance({addressOffset}, {resultOffset})");
+            var frame = VirtualMachine.ExecutionFrames.Peek() as WasmExecutionFrame
+                        ?? throw new InvalidOperationException("Cannot call GetMsgValue outside wasm frame");
```
Comment (soloway2010) [inline src/Lachain.Core/Blockchain/VM/ExternalHandler.cs:386]: Cannot call GetExternalBalance outside wasm frame

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs [hunk 2]
```diff
@@ -378,6 +378,53 @@ public static void Handler_Env_GetMsgValue(int dataOffset)
...
+
+            if (!result)
+                throw new InvalidContractException("Bad call to (get_external_balance)");
+        }
+        
+        public static void Handler_Env_GetBlockTimestamp(int dataOffset)
+        {
+            Logger.LogInformation($"Handler_Env_GetBlockTimestamp()");
+            var frame = VirtualMachine.ExecutionFrames.Peek() as WasmExecutionFrame
+                        ?? throw new InvalidOperationException("Cannot call GetMsgValue outside wasm frame");
```
Comment (soloway2010) [inline src/Lachain.Core/Blockchain/VM/ExternalHandler.cs:409]: Cannot call GetBlockTimestamp outside wasm frame

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs (hunk 1 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/VM/ExternalHandler.cs]: Rectified

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs (hunk 2 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/VM/ExternalHandler.cs]: Rectified

**Review by soloway2010 — APPROVED** [review]

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs [hunk 3]
```diff
@@ -378,6 +378,53 @@ public static void Handler_Env_GetMsgValue(int dataOffset)
...
+            // Get the TotalBlockHeight at the given Snapshot
+            var snapshot = frame.InvocationContext.Snapshot;
+            var blockHeight = snapshot.Blocks.GetTotalBlockHeight();
+            
+            // Get block at the given height
+            var block = snapshot.Blocks.GetBlockByHeight(blockHeight);
+            
+            // Get block's timestamp
+            if (block is null)
+                throw new InvalidContractException("Bad call to call function");
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/VM/ExternalHandler.cs:420]: I think it is better to have a more specific message here:
```suggestion
                throw new InvalidContractException("Bad call to (get_block_timestamp)");
```

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs [hunk 4]
```diff
@@ -378,6 +378,53 @@ public static void Handler_Env_GetMsgValue(int dataOffset)
...
+        {
+            Logger.LogInformation($"Handler_Env_GetExternalBalance({addressOffset}, {resultOffset})");
+            var frame = VirtualMachine.ExecutionFrames.Peek() as WasmExecutionFrame
+                        ?? throw new InvalidOperationException("Cannot call GetExternalBalance outside wasm frame");
+            
+            // Get the address from the given memory offset
+            var snapshot = frame.InvocationContext.Snapshot;
+            var addressBuffer = SafeCopyFromMemory(frame.Memory, addressOffset, 20);
+            if (addressBuffer is null)
+                throw new InvalidContractException("Bad call to call function");
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/VM/ExternalHandler.cs:392]: I think it is better to have more specific message here:
```suggestion
                throw new InvalidContractException("Bad call to (get_external_balance)");
```

**Review by sgladkov — CHANGES_REQUESTED** [review]

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs (hunk 3 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/VM/ExternalHandler.cs]: incorporated the comments

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs (hunk 4 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/VM/ExternalHandler.cs]: incorporated the comments

File: src/Lachain.Core/Blockchain/VM/ExternalHandler.cs [hunk 5]
```diff
@@ -378,6 +378,53 @@ public static void Handler_Env_GetMsgValue(int dataOffset)
...
+        {
+            Logger.LogInformation($"Handler_Env_GetExternalBalance({addressOffset}, {resultOffset})");
+            var frame = VirtualMachine.ExecutionFrames.Peek() as WasmExecutionFrame
+                        ?? throw new InvalidOperationException("Cannot call GetExternalBalance outside wasm frame");
+            
+            // Get the address from the given memory offset
+            var snapshot = frame.InvocationContext.Snapshot;
+            var addressBuffer = SafeCopyFromMemory(frame.Memory, addressOffset, 20);
+            if (addressBuffer is null)
+                throw new InvalidContractException("Bad call to (get_block_timestamp)");
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/VM/ExternalHandler.cs:392]: get_external_balance,  not  get_block_timestamp here.)
```suggestion
                throw new InvalidContractException("Bad call to (get_external_balance)");
```
You can accept this changes from the GitHub page directly,  by pressing the button "Commit suggestion" )

**Review by sgladkov — CHANGES_REQUESTED** [review]

**Review by sgladkov — APPROVED** [review]

**Pull request merged**